core/src/main/java/org/opensearch/sql/calcite/CalciteRelNodeVisitor.java

```diff
@@ -1692,21 +1652,61 @@ public RelNode visitStreamWindow(StreamWindow node, CalcitePlanContext context)
               .rowsTo(RexWindowBounds.CURRENT_ROW)
               .as(ROW_NUMBER_COLUMN_FOR_STREAMSTATS);
       context.relBuilder.projectPlus(streamSeq);
-    }
 
-    List<RexNode> overExpressions =
-        node.getWindowFunctionList().stream().map(w -> rexVisitor.analyze(w, context)).toList();
-    context.relBuilder.projectPlus(overExpressions);
-
-    // resort when there is by condition
-    if (hasGroup) {
+      // construct groupNotNull predicate: true only when all grouping fields are non-null
+      List<RexNode> groupByList =
+          groupList.stream().map(expr -> rexVisitor.analyze(expr, context)).toList();
+      List<RexNode> notNullList =
+          PlanUtils.getSelectColumns(groupByList).stream()
+              .map(context.relBuilder::field)
+              .map(context.relBuilder::isNotNull)
+              .toList();
+      RexNode groupNotNull =
+          notNullList.isEmpty()
+              ? context.relBuilder.literal(true)
+              : context.relBuilder.and(notNullList);
+
+      // get rawExpr
+      List<RexNode> overExpressions =
+          node.getWindowFunctionList().stream().map(w -> rexVisitor.analyze(w, context)).toList();
+      // wrap each expr: CASE WHEN groupNotNull THEN rawExpr ELSE CAST(NULL AS rawType) END
+      List<RexNode> wrappedOverExprs =
+          wrapWindowFunctionsWithGroupNotNull(overExpressions, groupNotNull, context);
+      context.relBuilder.projectPlus(wrappedOverExprs);
+      // resort when there is by condition
       context.relBuilder.sort(context.relBuilder.field(ROW_NUMBER_COLUMN_FOR_STREAMSTATS));
       context.relBuilder.projectExcept(context.relBuilder.field(ROW_NUMBER_COLUMN_FOR_STREAMSTATS));
+    } else {
+      List<RexNode> overExpressions =
+          node.getWindowFunctionList().stream().map(w -> rexVisitor.analyze(w, context)).toList();
+      context.relBuilder.projectPlus(overExpressions);
     }
-
     return context.relBuilder.peek();
   }
 
+  private List<RexNode> wrapWindowFunctionsWithGroupNotNull(
+      List<RexNode> overExpressions, RexNode groupNotNull, CalcitePlanContext context) {
+    List<RexNode> wrappedOverExprs = new ArrayList<>(overExpressions.size());
+    for (RexNode overExpr : overExpressions) {
+      RexNode rawExpr = overExpr;
+      String aliasName = null;
+      if (overExpr instanceof RexCall rc && rc.getOperator() == SqlStdOperatorTable.AS) {
+        rawExpr = rc.getOperands().get(0);
+        if (rc.getOperands().size() >= 2 && rc.getOperands().get(1) instanceof RexLiteral lit) {
+          aliasName = lit.getValueAs(String.class);
```

Comment:
Potential NPE: Add null check before calling `lit.getValueAs(String.class)`. The literal value could be null.

```java
  if (rc.getOperands().size() >= 2 && rc.getOperands().get(1) instanceof RexLiteral lit
      && lit.getValue() != null) {
      aliasName = lit.getValueAs(String.class);
  }
```

Reply:
`null instanceOf Class` always return false. So no need to check again.